src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,214 @@
+                    {
+                        map.Add(lineNum, (tagMappingSpan, changedSpan));
+                    }
+
+                    // If the map has a value and the value is not a syntax error, then rewrite the value in the map since
+                    // each line with errors will be ordered by importance
+                    if (value.mapTagSpan is not null && value.mapTagSpan.Tag.ErrorType is not PredefinedErrorTypeNames.SyntaxError)
+                    {
+                        map[lineNum] = (tagMappingSpan, changedSpan);
+                    }
+                }
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Iterates through the mapping of line number to span and draws the diagnostic in the appropriate position on the screen,
+        /// as well as adding the tag to the adornment layer.
+        /// </summary>
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            foreach (var (lineNum, (tagMappingSpan, snapshotSpan)) in map)
+            {
+                if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span))
+                {
+                    continue;
+                }
+
+                var geometry = viewLines.GetMarkerGeometry(span);
+                if (geometry != null)
+                {
+                    var tag = tagMappingSpan.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
+                    var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+
+                    // Need to get the SnapshotPoint to be able to get the IWpfTextViewLine
+                    var point = tagMappingSpan.Span.Start.GetPoint(snapshotSpan.Snapshot, PositionAffinity.Predecessor);
+                    if (point == null)
+                    {
+                        continue;
+                    }
+
+                    var lineView = TextView.GetTextViewLineContainingBufferPosition(point.Value);
+
+                    var visualElement = graphicsResult.VisualElement;
+                    if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfCode)
+                    {
+                        Canvas.SetLeft(visualElement, lineView.Right);
+                    }
+                    else if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                    {
+                        Canvas.SetLeft(visualElement, TextView.ViewportWidth - visualElement.DesiredSize.Width);
+                    }
```

Comment:
else throw unexpected value?

Reply:
I can throw an invalidoperationexception? I think that one fits best, please let me know if there is a more apt exception.